src/Microsoft.TemplateEngine.Orchestrator.RunnableProjects/Macros/BaseMacroConfig.cs

```diff
@@ -0,0 +1,214 @@
+            }
+            VariableName = variableName;
+            if (!string.IsNullOrWhiteSpace(dataType))
+            {
+                DataType = dataType!;
+            }
+        }
+
+        public string VariableName { get; }
+
+        public string Type { get; }
+
+        internal string DataType { get; } = "string";
+
+        internal abstract void Evaluate(IEngineEnvironmentSettings environmentSettings, IVariableCollection vars);
+    }
+
+    internal abstract class BaseMacroConfig<TMacro, TMacroConfig> : BaseMacroConfig, IMacroConfig
+        where TMacro : IMacro<TMacroConfig>
+        where TMacroConfig : BaseMacroConfig<TMacro, TMacroConfig>, IMacroConfig
+    {
+        protected BaseMacroConfig(TMacro macro, string variableName, string? dataType = null)
+            : base(macro.Type, variableName, dataType)
+        {
+            Macro = macro ?? throw new ArgumentNullException(nameof(macro));
+        }
+
+        internal TMacro Macro { get; }
+
+        internal override void Evaluate(IEngineEnvironmentSettings environmentSettings, IVariableCollection vars)
+        {
+            Macro.Evaluate(environmentSettings, vars, (TMacroConfig)this);
+        }
+
+        protected static string? GetOptionalParameterValue(IGeneratedSymbolConfig config, string parameterName, string? defaultValue = default)
+        {
+            return GetOptionalParameterValue(config, parameterName, ConvertJTokenToString, defaultValue);
+        }
+
+        protected static TVal? GetOptionalParameterValue<TVal>(IGeneratedSymbolConfig config, string parameterName, Func<string, IGeneratedSymbolConfig, string, TVal> converter, TVal? defaultValue = default)
+        {
+            if (config.Parameters.TryGetValue(parameterName, out string token))
+            {
+                return converter(token, config, parameterName);
+            }
+            return defaultValue;
+        }
+
+        protected static bool ConvertJTokenToBool(string token, IGeneratedSymbolConfig config, string parameterName)
+        {
+            try
+            {
+                var jToken = JToken.Parse(token);
+                if (jToken.Type is not JTokenType.Boolean and not JTokenType.String)
+                {
+                    throw new TemplateAuthoringException(string.Format(LocalizableStrings.MacroConfig_Exception_ValueShouldBeBoolean, config.VariableName, parameterName), config.VariableName);
+                }
+                if (bool.TryParse(jToken.ToString(), out bool result))
+                {
+                    return result;
+                }
+                throw new TemplateAuthoringException(string.Format(LocalizableStrings.MacroConfig_Exception_ValueShouldBeBoolean, config.VariableName, parameterName), config.VariableName);
+            }
+            catch (TemplateAuthoringException)
+            {
+                throw;
```

Comment:
Why catching then? (applies to below cases as well)

Reply:
changed to `when` clause